Hide devtools when closed, prevent focus

IanVS opened the pull request:
Fixes #2375.

Previously, opening and closing the devtools only changed it opacity (and subscribed/unsubscribed to queries).  That meant that it was still possible to tab through the elements in the panel, even when it was closed.

With this change, when the panel starts out closed, we set a `visibility: hidden` style, so that it is gone from both the accessibility tree and the tab order.  I've also set up transition listeners so that before a transition starts, if the devtools are opening, we turn visibility to visible, so that the opacity fade-in works correctly.  Likewise, on a transition end, we check if the panel is closed, and turn visibility to hidden.

I tested this by compiling, copying the built files into my project's node_modules, and running my app, and it worked as I expected.  The fade still occurs (both in and out), focus does not move into a closed devtools, and the devtools are hidden in the accessibility tree. 

This probably would have been more straightforward with a keyframe animation, but I am not sure how to do that using inline styles, or whether its even possible.

IanVS commented:
Oh cool, you can see it working in the codesandbox as well.  I haven't seen a github codesandbox bot before, that's pretty nifty!

TkDodo commented:
One of the best bots :)

TkDodo commented:
Thank you 👍 . I've tested this in the codesandbox, and it looks good. While testing, I found a semi-related issue that I probably introduced when I made the devtools unsubscribe when they are hidden:

- go to the codesandbox
- open the devtools
- see that you have the `posts` query there
- close the devtools
- navigate to one post
- open the devtools again
--> observe that you still only have one query (the `posts`) query in the panel
- trigger a refetch (e.g. window focus)
- second query appears

I think this happens because we put the queries into local state: 

https://github.com/tannerlinsley/react-query/blob/ff830db831f89ba7cebe66cf8c23ca218bdb3643/src/devtools/devtools.tsx#L339-L341

and then update in the effect when we are subscribed:

https://github.com/tannerlinsley/react-query/blob/ff830db831f89ba7cebe66cf8c23ca218bdb3643/src/devtools/devtools.tsx#L382

but since the update happens when we are closed, the component only re-renders when we open it and thus it still has the old queries in the local state.

---

one way to fix it is to really unmount when we close and mount again, because it would pick up the initial state of `useState` again. Another one would be to sync the state in a separate effect, calling `setUnsortedQueries` with the current value of what's in the query cache whenever `isOpen` changes. We could also pass `isOpen` as a `key` to the `ReactQueryDevtoolsPanel` which would make it unmount whenever that value changes. Not sure how that would work with the animation though, we'd have to try.

---

I understand if you don't want to tackle this at all, and we can open a separate issue as well. But if we decide to go for the unmount/remount solution because of this finding, I would rather not add the `visibility` change now just to remove it then. What do you think?

IanVS commented:
If we unmount/mount, we would lose other local state like filtering and such.  Is that acceptable?

If we want to keep that state, it sounds like your suggested approach would be to:

> sync the state in a separate effect, calling setUnsortedQueries with the current value of what's in the query cache whenever isOpen changes.

Are there any downsides to that approach?  It sounds relatively straightforward, and I can try it out here if you'd like.

TkDodo commented:
> It sounds relatively straightforward, and I can try it out here if you'd like.

cool, go ahead 😊

IanVS commented:
I believe I've fixed the stale queries, but just realized I've introduced a regression.  Opening the devtools and refreshing the page results in a `close` button but a missing panel.  I'll work on fixing that now...

IanVS commented:
Alright, I think I've taken care of that too.  Should be ready for another set of 👀 .  Thanks!

TkDodo commented on src/devtools/devtools.tsx:
```diff
@@ -378,9 +407,16 @@ export const ReactQueryDevtoolsPanel = React.forwardRef(
+        // So we'll manually populate our state
+        setUnsortedQueries(Object.values(queryCache.getAll()))
+
+        return () => {
+          unsubscribe()
+        }
```
would this also work?

```suggestion
        return unsubscribe
```

IanVS replied:
Yes, but it would do the same thing at the cost of another function declaration, right?

TkDodo replied:
quite the contrary. The previous solution declares a new inline arrow function and returns that, while the proposed solution just returns the already declared function _unsubscribe_ for react to invoke. It doesn't matter perf wise so we can also leave it as is :)

IanVS replied:
Oh sorry, I was reading this on my phone and it looked backwards for some reason, lol

TkDodo commented on src/devtools/devtools.tsx:
```diff
@@ -133,6 +133,33 @@ export function ReactQueryDevtools({
     setIsResolvedOpen(isOpen)
   }, [isOpen, isResolvedOpen, setIsResolvedOpen])
 
+  const handlePanelTransitionStart = React.useCallback(() => {
+    if (panelRef.current && isResolvedOpen) {
+      panelRef.current.style.visibility = 'visible'
+    }
+  }, [isResolvedOpen])
+
+  const handlePanelTransitionEnd = React.useCallback(() => {
+    if (panelRef.current && !isResolvedOpen) {
+      panelRef.current.style.visibility = 'hidden'
+    }
+  }, [isResolvedOpen])
```
If we move those function definitions _into_ the `useEffect` where they are used, we don't need to wrap them in `useCallback. What do you think?

IanVS replied:
Sure, I can do that.

TkDodo reviewed (approved):
great work, thanks 🚀

tannerlinsley commented:
:tada: This PR is included in version 3.17.2 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/react-query/v/3.17.2)
- [GitHub release](https://github.com/tannerlinsley/react-query/releases/tag/v3.17.2)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
